lib/web_ui/lib/src/engine/canvaskit/image_web_codecs.dart

```diff
@@ -302,3 +390,79 @@ bool isAvif(Uint8List data) {
   }
   return false;
 }
+
+Future<ByteData> readPixelsFromVideoFrame(VideoFrame videoFrame, ui.ImageByteFormat format) async {
+  if (format == ui.ImageByteFormat.png) {
+    final Uint8List png = await encodeVideoFrameAsPng(videoFrame);
+    return png.buffer.asByteData();
+  }
+
+  final ByteBuffer pixels = await readVideoFramePixelsUnmodified(videoFrame);
+
+  // Check if the pixels are already in the right format and if so, return the
+  // original pixels without modification.
+  if (_shouldReadPixelsUnmodified(videoFrame, format)) {
+    return pixels.asByteData();
+  }
+
+  // At this point we know we want to read unencoded pixels, and that the video
+  // frame is _not_ using the same format as the requested one.
+  final bool isBgrFrame = videoFrame.format == 'BGRA' || videoFrame.format == 'BGRX';
+  if (format == ui.ImageByteFormat.rawRgba && isBgrFrame) {
+    _bgrToRgba(pixels);
+    return pixels.asByteData();
+  }
+
+  // Last resort, just return the original pixels.
+  return pixels.asByteData();
+}
+
+/// Mutates the [pixels], converting them from BGRX/BGRA to RGBA.
+void _bgrToRgba(ByteBuffer pixels) {
+  final int pixelCount = pixels.lengthInBytes ~/ 4;
+  final Uint8List pixelBytes = pixels.asUint8List();
+  for (int i = 0; i < pixelCount; i += 4) {
+    // It seems even in little-endian machines the BGR_ pixels are encoded as
+    // big-endian, i.e. the blue byte is written into the lowest byte in the
+    // memory address space.
+    final int b = pixelBytes[i];
+    final int r = pixelBytes[i + 2];
+
+    // So far the codec has reported 255 for the X component, so there's no
+    // special treatment for alpha. This may need to change if we ever face
+    // codecs that do something different.
+    pixelBytes[i] = r;
+    pixelBytes[i + 2] = b;
+  }
+}
+
+bool _shouldReadPixelsUnmodified(VideoFrame videoFrame, ui.ImageByteFormat format) {
+  if (format == ui.ImageByteFormat.rawUnmodified) {
+    return true;
+  }
+
+  // Do not convert if the requested format is RGBA and the video frame is
+  // encoded as either RGBA or RGBX.
+  final bool isRgbFrame = videoFrame.format == 'RGBA' || videoFrame.format == 'RGBX';
+  return format == ui.ImageByteFormat.rawRgba && isRgbFrame;
+}
+
+Future<ByteBuffer> readVideoFramePixelsUnmodified(VideoFrame videoFrame) async {
+  final int size = videoFrame.allocationSize();
+  final Uint8List destination = Uint8List(size);
+  final JsPromise copyPromise = videoFrame.copyTo(destination);
+  await promiseToFuture<void>(copyPromise);
+  return destination.buffer;
+}
+
+Future<Uint8List> encodeVideoFrameAsPng(VideoFrame videoFrame) async {
+  final int width = videoFrame.displayWidth;
+  final int height = videoFrame.displayHeight;
+  final html.CanvasElement canvas = html.CanvasElement()
+    ..width = width
+    ..height = height;
+  final html.CanvasRenderingContext2D ctx = canvas.context2D;
+  drawVideoFrame(ctx, videoFrame, 0, 0);
```

Comment:
btw,  you can do:
```dart
ctx.drawImage(videoFrame, 0, 0);
```
but you probably have to make `VideoFrame implements CanvasImageSource`.

Reply:
Nice! Done.